### PR TITLE
First pass at Authentication summary in docs.

### PR DESCRIPTION
**Pull request by jgeewax**
Took a crack at this, related to https://github.com/GoogleCloudPlatform/gcloud-python/pull/805

### CODE REVIEW

**Comment by jgeewax** [comment]
@tmatsuo : FYI, feedback welcome.

File: docs/gcloud-auth.rst [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+the service account can be automatically discovered.
+If you're running your application elsewhere,
+you should download the service account JSON keyfile
+and point to it using an environment variable.
+
+-------------------
+Order of evaluation
+-------------------
+
+``gcloud`` will look in a bunch of different places
```
Comment (tmatsuo) [inline docs/gcloud-auth.rst]: I think the order is wrong according to:
https://developers.google.com/identity/protocols/application-default-credentials
env -> gcloud -> app engine -> compute engine

or do you do something to change the order?

Comment (jgeewax) [inline docs/gcloud-auth.rst]: OK - will swap.

File: docs/gcloud-auth.rst [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  :hidden:
+
+Authentication
+--------------
+
+=======
+Summary
+=======
+
+For the majority of cases,
```
Comment (tmatsuo) [inline docs/gcloud-auth.rst]: "For the majority of cases" doesn't help me determine whether I should rely on the library or not.

Can you add details as appendix like:

```
For the majority of cases (see [details](link to the appendix) for exceptions),
```

```
When you should create credentials on your own.
* blah
* blah
```

Comment (jgeewax) [inline docs/gcloud-auth.rst]: How about we make it clear "If you're running in GAE, or GCE or authenticated with the Cloud SDK, [authentication should 'just work']" ?

It should always work -- in the cases below we describe them -- but we want to say that in these situations you don't even have to think about it.

Comment (tmatsuo) [inline docs/gcloud-auth.rst]: or I would just say "Auth should just work for everyone. Let us know if it doesn't work for you." ;)

File: docs/gcloud-auth.rst [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+authentication should "just work"
+as credentials can be loaded implicitly
+based on the environment you're application is running in
+and therefore there is no extra code required to authenticate.
```
Comment (tmatsuo) [inline docs/gcloud-auth.rst]: I think we should add something like
```
Note that for certain environments there are additional configuration steps needed, refer to the section below that explains other needed steps for your environment.
```

Those steps are:
* For compute engine, you have to create the instance with appropriate scopes.
* Create service account, download json, etc

I would also add that, in any cases, the service account should have appropriate permissions on your API. (FYI, some old app engine apps don't have the app engine's service account in the permission list. Also, you can delete those default service accounts from the permission list manually.)

File: docs/gcloud-auth.rst (hunk 2 above)
Comment (tmatsuo) [inline docs/gcloud-auth.rst]: SG
What about the service account? There are additional steps needed, but the same code still works.

Do you think of any cases that developers need to explicitly create credentials?

Comment (jgeewax) [inline docs/gcloud-auth.rst]: This feels like something that would be better linked off to. Maybe we can add a link to...

- https://developers.google.com/identity/protocols/OAuth2ServiceAccount or
- https://cloud.google.com/storage/docs/authentication#service_accounts ?

I like the second one better, but it's under the bucket of "Cloud Storage" so could possibly confuse people... (But it has screen shots... which is nice...)

I'm going to add the second one -- feel free to change if necessary.

File: docs/gcloud-auth.rst (hunk 3 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: This is a tricky one. We can't cover *everything* under the sun, but I agree that these are serious potential pitfalls.

How about we add a FAQ at the end that has a few things:

1. How do I create a service account? (see links above)
1. I'm on GCE, but when I get an Access Denied message when trying to talk to Datastore (or Storage, or whatever). What gives?

(And I'll let you open a PR for more...)

Comment (tmatsuo) [inline docs/gcloud-auth.rst]: SG

**Comment by jgeewax** [comment]
@dhermes can you take a look?

File: docs/gcloud-auth.rst [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+... using a JSON keyfile
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: python
+
+   from gcloud.credentials import get_for_service_account_json
+   from gcloud import pubsub
+
+   # Create the credentials from the keyfile and set the default connection.
+   credentials = get_for_service_account_json('/path/to/key.json')
```
Comment (jgeewax) [inline docs/gcloud-auth.rst]: @dhermes Do I need to do some magic with `.create_scoped(...)` here? Or is this correct and the `pubsub.Connection.__init__` method will handle setting the proper scope..?

File: docs/gcloud-auth.rst (hunk 1 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: Not that isn't the order it actually happens in. See [the code][1].

The `cls._implicit_credentials_from_files` method [checks][2] first the `GOOGLE_APPLICATION_CREDENTIALS` environment variable and then the default credentials created by `gcloud auth login`.

[1]: https://github.com/google/oauth2client/blob/22a532de1d632a98beca7323a66a01fa894fbb86/oauth2client/client.py#L1159-L1163
[2]: https://github.com/google/oauth2client/blob/22a532de1d632a98beca7323a66a01fa894fbb86/oauth2client/client.py#L1120-L1122

File: docs/gcloud-auth.rst [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+**If you're running in Compute Engine or App Engine**,
+authentication should "just work"
+as credentials can be loaded automatically
+and therefore there is no extra code required to authenticate.
+
+**If you're developing locally**,
+the easiest way to authenticate is using the Cloud SDK.
+
+**If you're running your application elsewhere**,
+you should download the service account JSON keyfile
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: s/the service account JSON keyfile/a service account JSON keyfile

Using "the" makes it sound like there is just one.

File: docs/gcloud-auth.rst [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+and point to it using an environment variable.
+
+-----------------------------------------------
+In what order does gcloud discover credentials?
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: The `gcloud` name is pretty ambiguous and it confuses people. Our project is `gcloud-python`, we are `gcloud` on PyPI and obviously there is the `gcloud` CLI tool.

File: docs/gcloud-auth.rst [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+-----------------------------------------------
+
+``gcloud`` will look in a bunch of different places
+for authentication credentials, in the following order:
+
+#. Credentials you explicitly set in code
+#. Credentials from an environment variable (``GOOGLE_APPLICATION_CREDENTIALS``)
+#. Credentials provided by the Cloud SDK (``gcloud auth login``)
+#. Credentials provided when running in Google App Engine
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: As mentioned, App Engine comes before environment variables.

Though committing to any order is arbitrary / unnecessary.

File: docs/gcloud-auth.rst [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+#. Credentials provided when running in Google Compute Engine
+
+This means that you can set your credentials a bunch of different ways.
+We recommend using some form of implicitly discovered credentials
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: I don't think this is a broad recommendation for all devs.

People writing multithreaded code may as well find it's easier to deal with concurrent state by passing explicit values.

File: docs/gcloud-auth.rst [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+(ie, set in an environment variable or auto-discovered in GCE/GAE)
+so that your code can be written the same and
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: Suggest changing

> so that your code can be written the same and simply run in the environment of your choice.

to

> so that your code can be simply run in the environment of your choice with no changes.

File: docs/gcloud-auth.rst [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+#. Download and install the Cloud SDK (You can get the Cloud SDK at https://cloud.google.com/sdk)
+#. Authenticate using OAuth2 (``gcloud auth login``)
+#. Run your code (ie, ``python myscript.py``)
+
+
+After you do this,
+your script will look for the authentication token
+that you got when authenticating,
+and you won't need to think about authentication in your code at all.
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: I'd be more comfortable if there was a strong emphasis that this isn't recommended suitable for production code.

File: docs/gcloud-auth.rst [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+
+This means your code doesn't have to change at all,
+and can run with different credentials depending on the environment.
+
+Here's what this looks like:
+
+.. code-block:: bash
+
+   $ export GOOGLE_APPLICATION_CREDENTIALS="/path/to/key.json"
+   $ python
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: If you like, these line's can be condensed into one

```
$ env GOOGLE_APPLICATION_CREDENTIALS="/path/to/key.json" python
```

File: docs/gcloud-auth.rst [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+   >>> from gcloud import pubsub
+   >>> pubsub.Topic('topic_name').create()
+
+--------------------------------------
+Setting the keyfile explicitly in code
+--------------------------------------
+
+If you really want to set a specific keyfile in code
+(maybe for a code snippet to send to a friend?)
+you can do this, but it's really not recommended.
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: This language strongly commits to an opinion I didn't know we had.

Some (many?) developers prefer declarative defaults and constants that they can reason about directly. Our "it's implicit, we got you covered" code makes many devs (@tseaver included) worry they don't know how all the moving pieces fit together.

File: docs/gcloud-auth.rst [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+
+~~~~~~~~~~~~~~~~~~~~~~~~
+... using a JSON keyfile
+~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: The use of `~~~` becomes an `<h4>`, which has no CSS indicating it is different from other text. 

Any reason not to use an `<h3>`?

File: docs/gcloud-auth.rst [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+       '/path/to/key.p12')
+   connection = pubsub.Connection(credentials=credentials)
+   pubsub.set_default_connection(connection)
+
+   # Now you can interact with the service as usual.
+   pubsub.Topic('topic_name').create()
+
+=====================================
+Authenticating from inside GCE or GAE
+=====================================
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: Any issue to axe the declarative snippets I included in #805?

They are quite simple:

```python
from oauth2client.gce import AppAssertionCredentials
credentials = AppAssertionCredentials([])
# AND
from oauth2client.appengine import AppAssertionCredentials
credentials = AppAssertionCredentials([])
```

File: docs/gcloud-auth.rst (hunk 4 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: At current, you need the magic, but that shouldn't be the case. Will send a PR.

File: docs/gcloud-auth.rst [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+use a different Service Account.
+In that case, jump up to the section about
+using a specific keyfile.
+Thanks to the order of evaluation,
+the keyfile you specify explicitly or via the environment variable
+will take precedence over the automatically discovered key in the
+App Engine or Compute Engine environment.
+
+=============================================
+Using multiple credentials in the same script
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: s/same script/same code

File: docs/gcloud-auth.rst [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+=============================================
+
+There may be times where you want to use
+multiple different sets of credentials inside the same script.
+To do this, you should create miltiple connections
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: s/miltiple/multiple

File: docs/gcloud-auth.rst [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+
+   pubsub.Topic('topic1').create()
+   pubsub.Topic('topic2', project='project2', connection=connection2).create()
+
+===============
+Troubleshooting
+===============
+
+--------------------------------------------------
+You keep mentioning key.json... How do I get that?
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: 1. This isn't troubleshooting, it's just another step they need to complete.
1. Can we make this just a tad bit less conversational?

File: docs/gcloud-auth.rst [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+--------------------------------------------------
+
+When we say ``key.json``
+we're referring to what we call a "Service Account".
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: Not quite. `key.json` refers to credentials for a service account, not to the account itself (I realize you know the difference, just the text doesn't quite work).

File: docs/gcloud-auth.rst [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+These are sort of like passwords
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: We're talking to devs, saying "credentials" should be sufficient. They are much fancier than passwords. Though when describing to my grandma, I'd probably say password. (But again, not grandma, devs.)

File: docs/gcloud-auth.rst [hunk 20]
```diff
@@ -0,0 +1,264 @@
...
+
+Service Accounts are pretty easy to create,
+just head over to the `Credentials`_ page
+in the `Google Developers Console`_,
+create a new Client ID,
+and generate a new JSON key.
+
+If you're a bit confused,
+take a look at the step by step guide here:
+https://cloud.google.com/storage/docs/authentication#service_accounts
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: Big +1 on this link

File: docs/gcloud-auth.rst [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+I'm still having trouble. Help?
+-------------------------------
+
+If you're having trouble authenticating,
+feel free to open a `Github Issue`_ on this project
+and hopefully we can help
+(and add the answer to your question here).
+
+.. _Google Developers Console: https://console.developers.google.com/project
+.. _Credentials: https://pantheon.corp.google.com/project/_/apiui/credential
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: Check yoself before you wreck yoself with this link

File: docs/gcloud-auth.rst [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+.. _Github Issue: https://github.com/GoogleCloudPlatform/gcloud-python/issues/new?title=I+need+help+with+auth!
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: Can we use something without an exclamation point? How about

https://github.com/GoogleCloudPlatform/gcloud-python/issues/new?title=Authentication+Question%3A

File: docs/gcloud-auth.rst [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+(at the top of the screen, towards the right),
+scroll down to the section called "Project Access",
+and set the appropriate access
+for the service you want your VM to have
+(ie, Datastore -> Enabled).
+
+.. note::
+
+   Unfortunately,
+   you can't change these settings after the VM is created. :(
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: Can you get rid of the frowny face?

File: docs/gcloud-auth.rst [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+in the `Google Developers Console`_,
+create a new Client ID,
+and generate a new JSON key.
+
+If you're a bit confused,
+take a look at the step by step guide here:
+https://cloud.google.com/storage/docs/authentication#service_accounts
+
+----------------------------------------------------------------
+I'm running in Compute Engine, but can't access Datastore. Help?
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: We should make this more generic. We can include a stacktrace with a 403 error and just talk about turning services on in a project (getting a VM custom scopes is still a separate issue).

File: docs/gcloud-auth.rst (hunk 12 above)
Comment (tseaver) [inline docs/gcloud-auth.rst]: In particular, "majyk" defaults create steep learning curves:  once you get past the point where they are the only obvious thing, they make figuring out how to do the Right Thing(TM) harder.  For instance, consider a push-to-deploy webapp:  it is a goal to have *all* configuration variables explicitly managed inside VCS-managed files.  Reading such a config file and configuring the path for the keyfile from an entry in it is a logical way to accomplish that requirement.

Comment (jgeewax) [inline docs/gcloud-auth.rst]: I tend to agree with you that it can be confusing to have lots of magic (hence why I really really want these options described here).

What would the ideal auth story look like if you guys could pick anything you want? (code) @dhermes @tseaver

File: docs/gcloud-auth.rst (hunk 13 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Hierarchy is:

Title -> Auth -> Authenticating with a specific keyfile -> Setting explicitly in code -> ... Using a JSON keyfile

Wouldn't that mean h4 is the right tag?

Comment (dhermes) [inline docs/gcloud-auth.rst]: Yeah my "any reason not to use an `<h3>`" doesn't make much sense.

However, the fact that there is no CSS to distinguish our `<h4>` tags is still an issue.

**Comment by tmatsuo** [comment]
I came up with a little concern about thread safety and multiprocessing safety.

Is the implicit method thread safe?
Also, trickier thing is that, even if your client object is thread safe, it could break with multiprocessing.

For example, when I did the following, the apiary client object breaks:

1. create a client
2. issue an API call in the main process
3. fork sharing the client
4. issue an API with the shared client in the child process

BTW the document looks great now :)

File: docs/gcloud-auth.rst (hunk 8 above)
Comment (tmatsuo) [inline docs/gcloud-auth.rst]: The credentials are thread safe. Only httplib2.Http object is the problem. It's easily avoidable with using thread local Http object.
However, if you use multiprocessing, this approach sometimes breaks. (Maybe there is a workaround)

Anyways, is it possible for gcloud to always use thread local Http object? Then multithreaded code authors can just use the implicit method.

**Comment by dhermes** [comment]
@tmatsuo Thanks for pointing this out. I've also worried about how our implicit code would work in a multithreaded environment. (Easy becomes hard very quickly.)

----

Are you referring to our library or `google-api-python-client`? Also, can you send sample code and stacktraces if it is us breaking.

**Comment by tmatsuo** [comment]
@dhermes That's the `apiary` client, so yes, it was `google-api-python-client`. However, if you're using httplib2.Http object, it could happen to anyone I guess.

Unfortunately I lost the stacktraces now, I tried to reproduce it, but now I'm failing to repro.
Basically what I did was
* store the apiary client object in a thread local variable
* use that client in the main process
* fork with multiprocessing (thus the client object is copied to the child)
* use the client in the child process

Then it threw some kind of SSL handshake error.

I will update here once I can reproduce it.

**Comment by dhermes** [comment]
Thanks. That's "good news" for us since we don't do config with `httplib2` (we use raw `httplib` / `http.client` to check compute engine for config).

The flipside is that this could be a larger issue in a threaded environment. I thought I filed an issue about enabling a "bring your own `http`" but couldn't track it down. (We don't necessarily need to pigeonhole people into using `httplib2`.)

File: docs/gcloud-auth.rst (hunk 22 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Sure - updated.

(I couldn't find a way... fiddled with the URL and looked in the docs... Maybe we can ask?)

File: docs/gcloud-auth.rst (hunk 21 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Fixed.

File: docs/gcloud-auth.rst (hunk 23 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Done.

File: docs/gcloud-auth.rst (hunk 18 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Fixed - thanks.

File: docs/gcloud-auth.rst (hunk 16 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Fixed.

File: docs/gcloud-auth.rst (hunk 15 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Fixed.

File: docs/gcloud-auth.rst (hunk 4 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Cool - will leave this as is then.

File: docs/gcloud-auth.rst (hunk 13 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Gotcha, we should open an issue to update some CSS for this. Will leave as is for now though.

File: docs/gcloud-auth.rst (hunk 10 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Added a note.

File: docs/gcloud-auth.rst (hunk 9 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Done.

File: docs/gcloud-auth.rst (hunk 7 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Moved GAE up.

File: docs/gcloud-auth.rst (hunk 6 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Updated to gcloud-python in both places.

File: docs/gcloud-auth.rst (hunk 5 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Gotcha, fixed.

File: docs/gcloud-auth.rst [hunk 25]
```diff
@@ -0,0 +1,271 @@
...
+#. Credentials you explicitly set in code
+#. Credentials provided when running in Google App Engine
+#. Credentials from an environment variable (``GOOGLE_APPLICATION_CREDENTIALS``)
+#. Credentials provided by the Cloud SDK (``gcloud auth login``)
+#. Credentials provided when running in Google Compute Engine
+
+This means that you can set your credentials a bunch of different ways.
+We recommend using some form of implicitly discovered credentials
+(ie, set in an environment variable or auto-discovered in GCE/GAE)
+so that youre code can be
```
Comment (dhermes) [inline docs/gcloud-auth.rst]: s/youre/your

Comment (jgeewax) [inline docs/gcloud-auth.rst]: Done. :-/

File: docs/gcloud-auth.rst (hunk 17 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Sure -- let's do that in a separate request to get this pushed ? Or do you think that's a show-stopper?

File: docs/gcloud-auth.rst (hunk 22 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: No worries.

File: docs/gcloud-auth.rst (hunk 17 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: Not a show-stopper, but I have plenty of comments still pending, so you will likely be making other commits.

(Many of the comments are hidden as "outdated diff", I can repost them if you like to make it easier to see.)

File: docs/gcloud-auth.rst (hunk 19 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: A lot of people are so used to passwords (or AWS's ACCESS_KEY_ID and SECRET_KEY) which is why I'm hand-holding a bit more here... Particularly MySQL or S3 users...

File: docs/gcloud-auth.rst (hunk 8 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: So what's the verdict here? Any change needed?

Comment (dhermes) [inline docs/gcloud-auth.rst]: We still don't have to recommend it, even if it is supported. We certainly shouldn't recommend it without testing it.

File: docs/gcloud-auth.rst (hunk 12 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: I believe we'd be happier (and have simpler code) if we required people to just call

```python
<package>.set_defaults()
```

before any requests will work.

File: docs/gcloud-auth.rst (hunk 13 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: #846

File: docs/gcloud-auth.rst (hunk 19 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: Do we really think someone who has configured a MySQL DB needs handholding?

File: docs/gcloud-auth.rst (hunk 12 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: I think we need to take that discussion elsewhere as I suspect it will make this thread super long. Want to open an issue for that?

File: docs/gcloud-auth.rst (hunk 8 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: So can we articulate why we would not recommend auto-discovered credentials (ie, using an environment variable) over explicitly defined credentials (ie, putting a file path in code)?

And did we not test this?

File: docs/gcloud-auth.rst (hunk 12 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: Sure (was just trying to answer your question about our ideal).

Filed #847

File: docs/gcloud-auth.rst (hunk 19 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: I think I may have implied that people using Google Cloud are the same people who have configured a MySQL Database. I meant "turned on".

We also have to consider the scenario where people are coming from App Engine which requires **zero** authentication to talk to Datastore...

I think it's worthwhile to clarify what the thing is with one sentence. If you want to shorten it up or correct it, we should do that. It's my feeling that if someone can't figure out what `key.json` is on their own, they need hand holding and probably don't even know what a service accounts (or credentials JSON files) are...

File: docs/gcloud-auth.rst (hunk 8 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: As mentioned above

> People writing multithreaded code may as well find it's easier to deal with concurrent state by passing explicit values.

I'm not saying we don't test implicit vs. explicit. I'm saying we don't test in a multithreaded environment.

**Comment by dhermes** [comment]
FYI @jgeewax @tseaver et al I just discovered a quick way to expand all outdated comments:

```javascript
$(".outdated-diff-comment-container").addClass("open")
```

from the JS console

File: docs/gcloud-auth.rst (hunk 19 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: It's fine to handhold, I just don't think we need to introduce a false/incomplete analogy to a password.

How about changing

> These are sort of like passwords which let you access the different cloud services.

to

> A service account is like a user account for a machine (rather than a human) and
the `key.json` file allows the account to authenticate.

Beyond that, the identity folks at Google should have a document we can point them to.

File: docs/gcloud-auth.rst (hunk 7 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Just to confirm, the order should be:

1. Explicit in code
1. GAE
1. environment variable
1. gcloud auth login file
1. GCE

According to: https://github.com/google/oauth2client/blob/22a532de1d632a98beca7323a66a01fa894fbb86/oauth2client/client.py#L1159-L1163

File: docs/gcloud-auth.rst (hunk 11 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Leaving this as is for now, can always change later.

File: docs/gcloud-auth.rst (hunk 14 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Not sure what you mean here. What should be changed?

File: docs/gcloud-auth.rst (hunk 19 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Updated.

File: docs/gcloud-auth.rst (hunk 24 above)
Comment (jgeewax) [inline docs/gcloud-auth.rst]: Not sure exactly what you mean. Maybe we can put that into a separate PR.

File: docs/gcloud-auth.rst (hunk 14 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: Those snippets were included in my doc, but aren't here.

Is there a reason you don't want to include them for the folks that want to be declarative?

File: docs/gcloud-auth.rst (hunk 7 above)
Comment (dhermes) [inline docs/gcloud-auth.rst]: Correct

**Comment by dhermes** [comment]
@jgeewax I just tried to revisit this and #805, but realized it's not worth moving forward on until the client pattern is fully in.

When I do move forward on it, I'm going to start with a tiny document (i.e. one paragraph) and then "launch and iterate" so we don't have so many comments to address in a single PR.

**Comment by dhermes** [comment]
Though you may notice that this little review spawned #950 and #951.
